File: FeatureSelection.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <sstream>
#include <algorithm>
#include <cmath>
#include <chrono>

using namespace std;

vector<double> getClassLabels(vector<vector<double>> data){
    string line, classLabel;
    vector<double> classLabels;
    for (unsigned int i = 0; i < data.size(); ++i) {
        classLabels.push_back(data[i][0]); //Only push back first element for each row which is the class label
    }
    return classLabels;
}

vector<vector<double>> readData(string fileName){
    string line, featureValue;
    vector<double> currRow;
    vector<vector<double>> fullData;
    ifstream file (fileName);
    while (getline(file,line)) { //Loop through every row
        stringstream ss(line);
        while (ss >> featureValue) { //Add row values into currRow
            currRow.push_back(stod(featureValue)); //Add curr rows into fullData as doubles
        }
        fullData.push_back(currRow);
        currRow.clear(); //Clear currRow to get ready for next row's data
    }
    file.close();
    return fullData;
}

vector<vector<double>> editData(vector<vector<double>> data){
    for (unsigned int i = 0; i < data.size(); ++i) { //Remove class label from every row so only features in the row remains
        data[i].erase(data[i].begin()); 
    }
    return data;
}

vector<vector<double>> removeFeatures(vector<vector<double>> featureData, vector<int> classSet, int featureToAdd){
    vector<int> columnsToKeep = classSet;
    columnsToKeep.push_back(featureToAdd); //Columns we want to keep, 0 out every other column
    for (unsigned int i = 0; i < featureData.size(); ++i) {
        for (unsigned int j = 0; j < featureData[i].size(); ++j) {
            if (find(columnsToKeep.begin(), columnsToKeep.end(), j) == columnsToKeep.end()){
                featureData[i][j] = 0; //Set columns not in classSet and featureToAdd to 0
            }
        }
    }
    return featureData;
}

vector<vector<double>> removeFromSet(vector<vector<double>> featureData, vector<int> classSet, int featureToRemove){
    vector<int> columnsToKeep = classSet;
    columnsToKeep.erase(remove(columnsToKeep.begin(),columnsToKeep.end(),featureToRemove),columnsToKeep.end()); 
    for (unsigned int i = 0; i < featureData.size(); ++i) {
        for (unsigned int j = 0; j < featureData[i].size(); ++j) {
            if (find(columnsToKeep.begin(), columnsToKeep.end(), j) == columnsToKeep.end()){
                featureData[i][j] = 0; //Set columns not in classSet to 0
            }
        }
    }
    return featureData;
}

double crossValidation(vector<vector<double>> featureData, vector<int> currentSet, int featureToAddOrRemove, vector<double> classLabels, int userNum){
    if (userNum == 1) //Forward Search
        featureData = removeFeatures(featureData, currentSet, featureToAddOrRemove);
    else if (userNum == 2) //Backward Elimination
        featureData = removeFromSet(featureData, currentSet, featureToAddOrRemove);

    double accuracy = 0.0, distance = 0.0, classLabelToClassify, nearestNeighborDistance, nearestNeighborLocation, nearestNeighborLabel, numCorrectlyClassified = 0.0;
    vector <double> objectToClassify, objectToCompare;

    for (unsigned int i = 0; i < featureData.size(); ++i) {
        objectToClassify = featureData[i];
        classLabelToClassify = classLabels[i];
        nearestNeighborDistance = 4294967295; //Some large value for initialization
        nearestNeighborLocation = 4294967295;

        for (unsigned int k = 0; k < featureData.size(); ++k) {
            if (k != i) { //Don't compare same feature to itself
                distance = 0.0;
                objectToCompare = featureData[k];
                for (unsigned int l = 0; l < objectToClassify.size(); ++l) {
                    distance += pow(objectToClassify[l] - objectToCompare[l],2);
                }
                distance = sqrt(distance); //Calculate distance
                if (distance < nearestNeighborDistance) { //Update nearestNeighborDistance, nearestNeighborLocation, and nearestNeighborLabel
                    nearestNeighborDistance = distance;
                    nearestNeighborLocation = k;
                    nearestNeighborLabel = classLabels[k];
                }
            }
        }
        if (classLabelToClassify == nearestNeighborLabel) { //If correctly classified
            numCorrectlyClassified++;
        }
    }
    accuracy = numCorrectlyClassified / featureData.size();
    return accuracy;
}

void forwardFeatureSearch(vector<vector<double>> featureData, vector<double> classLabels, int userNum){
    vector<int> currentSetOfFeatures, bestFeatures;
    double accuracy = 0.0, bestSoFarAccuracy = 0.0, bestTotalAccuracy = 0.0;
    
    for (unsigned int i = 0; i < featureData[0].size(); ++i) {
        int featureToAddAtCurrentLevel = -1, bestFeatureAtLevel = -1;
        bestSoFarAccuracy = 0.0;
        cout << "On level " << i + 1 << " of the search tree" << endl;

        for (unsigned int k = 0; k < featureData[0].size(); ++k) { //Loop through every feature (column)
            if (find(currentSetOfFeatures.begin(),currentSetOfFeatures.end(),k) == currentSetOfFeatures.end()){ // If the value is not in the current set of features
                accuracy = crossValidation(featureData,currentSetOfFeatures,k,classLabels, userNum);
                cout << "--Testing adding feature " << k + 1 << " with current accuracy of: " << accuracy << " using feature(s) {";
                for (unsigned int i = 0; i < currentSetOfFeatures.size(); ++i) {
                    cout << currentSetOfFeatures[i] + 1 << ", ";
                }
                cout << k + 1 << "} " << endl;
                if (accuracy > bestSoFarAccuracy){ //Update bestSoFarAcccuracy
                    bestSoFarAccuracy = accuracy;
                    featureToAddAtCurrentLevel = k;
                }
                if (accuracy > bestTotalAccuracy){ //Update bestTotalAccuracy
                    bestTotalAccuracy = accuracy;
                    bestFeatureAtLevel = k;
                }
            }
        }
        cout << "On level " << i + 1 << ", I added feature " << featureToAddAtCurrentLevel + 1 << " with accuracy of " << bestSoFarAccuracy << endl;
        currentSetOfFeatures.push_back(featureToAddAtCurrentLevel);
        if (bestFeatureAtLevel >= 0){
            bestFeatures.push_back(bestFeatureAtLevel);
        }
        bestFeatureAtLevel = -1;
    }
    cout << "------------------------------------------------------------" << endl;
    cout << "Best features to use: " << endl;
    for (unsigned int i = 0; i < bestFeatures.size(); ++i){ 
        cout << bestFeatures[i] + 1 << " " ;
    }
    cout << endl << "The best accuracy from this data set is: " << bestTotalAccuracy << endl;
}

void backwardFeatureSearch(vector<vector<double>> featureData, vector<double> classLabels, int userNum){
    vector<int> currentSetOfFeatures, bestFeatures, printFeatures;
    double accuracy = 0, bestSoFarAccuracy = 0.0, bestTotalAccuracy = 0.0;

    for (unsigned int i = 0; i < featureData[0].size(); ++i) { //Start with every feature 
        currentSetOfFeatures.push_back(i);
    }
    
    for (unsigned int i = 0; i < featureData[0].size(); ++i) {
        int featureToRemoveAtCurrentLevel = -1;
        bestSoFarAccuracy = 0.0;
        cout << "On level " << i + 1 << " of the search tree" << endl;

        for (unsigned int k = 0; k < featureData[0].size(); ++k) {
            if (find(currentSetOfFeatures.begin(),currentSetOfFeatures.end(),k) != currentSetOfFeatures.end()){ // If the value is in the current set of features
                accuracy = crossValidation(featureData,currentSetOfFeatures,k,classLabels, userNum);
                cout << "--Testing removing feature " << k + 1 << " with current accuracy of: " << accuracy <<  " after removing the feature from the set {";
                printFeatures = currentSetOfFeatures;
                printFeatures.erase(remove(printFeatures.begin(),printFeatures.end(),k),printFeatures.end());
                for (unsigned int i = 0; i < printFeatures.size(); ++i) {
                    if (i < printFeatures.size() - 1)
                        cout << printFeatures[i] + 1 << ", ";
                    else    
                        cout << printFeatures[i] + 1;
                }
                cout << "}" << endl;
                if (accuracy > bestSoFarAccuracy){ //Update bestSoFarAccuracy
                    bestSoFarAccuracy = accuracy;
                    featureToRemoveAtCurrentLevel = k;
                }
                if (accuracy > bestTotalAccuracy){ //Update bestTotalAccuracy
                    bestTotalAccuracy = accuracy;
                    bestFeatures = currentSetOfFeatures;
                    bestFeatures.erase(remove(bestFeatures.begin(),bestFeatures.end(),k),bestFeatures.end());
                }
            }
        }
        cout << "On level " << i + 1 << ", I removed feature " << featureToRemoveAtCurrentLevel + 1 << endl;
        currentSetOfFeatures.erase(remove(currentSetOfFeatures.begin(),currentSetOfFeatures.end(),featureToRemoveAtCurrentLevel),currentSetOfFeatures.end()); //Remove the feature
    }
    cout << "------------------------------------------------------------" << endl;
    cout << "Best features to use: " << endl;
    for (unsigned int i = 0; i < bestFeatures.size(); ++i){ 
        cout << bestFeatures[i] + 1 << " " ;
    }
    cout << endl << "The best accuracy from this data set is: " << bestTotalAccuracy << endl;
}

int main(){
    string fileName;
    int algorithmNum;

    cout << "Welcome to Nathan's Feature Selection Algorithm." << endl;
    cout << "Type in the name of the file to test: " << endl;
    cin >> fileName;
    cout << "Type the number of the algorithm you want to run." << endl;
    cout << "1. Forward Selection" << endl << "2. Backward Elimination" << endl;
    cin >> algorithmNum;

    vector<vector<double>> data = readData(fileName); //2 dimensional vector containing row x column
    vector<double> classLabels = getClassLabels(data); //Vector containing only class labels
    vector<vector<double>> dataMinusClass = editData(data); //2 dimensional vector containing only features

    cout << "This dataset has " << dataMinusClass[0].size() << " features (not including the class attribute), with " << dataMinusClass.size() << " instances." << endl;

    if (algorithmNum == 1) {
        cout << "Beginning Search" << endl;
        auto start = std::chrono::high_resolution_clock::now();
        forwardFeatureSearch(dataMinusClass, classLabels, algorithmNum);
        auto finish = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(finish - start);
        cout << "Duration: " << duration.count() << " milliseconds" << endl;
    }
    else if (algorithmNum == 2) {
        cout << "Beginning Search" << endl;
        auto start = std::chrono::high_resolution_clock::now();
        backwardFeatureSearch(dataMinusClass, classLabels, algorithmNum);
        auto finish = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(finish - start);
        cout << "Duration: " << duration.count() << " milliseconds" << endl;
    }
    else 
        cout << "Invalid input. Exiting" << endl;

    return 0;
}

File: FeatureSelection.py
import numpy as np
import copy
from math import sqrt

def getLabels(data):
    classLabels = []
    for i in range(len(data)):
        classLabels.append(data[i][0])
    return classLabels

def editData(data):
    for i in range(len(data)):
        del data[i][0]
    return data

def removeFeatures(data, current_set, feature_to_add):
    columns_to_keep = copy.deepcopy(current_set)
    columns_to_keep.append(feature_to_add)
    for i in range(len(data)):
        for j in range(len(data[0])):
            if j not in columns_to_keep:
                data[i][j] = 0
    return data

def removeFromSet(data, current_set, feature_to_remove):
    columns_to_keep = copy.deepcopy(current_set)
    columns_to_keep.remove(feature_to_remove)
    for i in range(len(data)):
        for j in range(len(data[0])):
            if j not in columns_to_keep:
                data[i][j] = 0
    return data    

def crossValidation(datas,current_sets,feature_to_adds, class_label, userNum):
    data = copy.deepcopy(datas)        
    current_set = copy.deepcopy(current_sets)
    feature_to_add = copy.deepcopy(feature_to_adds)
    class_labels = copy.deepcopy(class_label)
    if userNum == "1":
        data = removeFeatures(data,current_set,feature_to_add)
    elif userNum == "2":
        data = removeFromSet(data,current_set,feature_to_add)
    distance = 0.0
    num_correctly_classified = 0
    object_to_classify = []
    object_to_compare = []
    for i in range(len(data)):
        object_to_classify = data[i]
        class_label_to_classify = class_labels[i]
        nearest_neighbor_distance = 99999
        for j in range(len(data)):
            if j != i:
                distance = 0.0
                object_to_compare = data[j]
                for l in range(len(object_to_classify)):
                    distance = distance + pow(object_to_classify[l] - object_to_compare[l],2)
                distance = sqrt(distance)
                if distance < nearest_neighbor_distance:
                    nearest_neighbor_distance = distance
                    nearest_neighbor_label = class_labels[j]
        if class_label_to_classify == nearest_neighbor_label:
            num_correctly_classified = num_correctly_classified + 1
    accuracy = num_correctly_classified / len(data)
    return accuracy
                
def forwardSelection(data, class_labels, userNum):
    class_label = copy.deepcopy(class_labels)
    current_set_of_features = []
    best_features = []
    best_total_accuracy = 0.0
    accuracy = 0.0
    best_so_far_accuracy = 0.0
    for i in range(len(data[0])): 
        feature_to_add_at_current_level = -1
        best_feature_at_level = -1
        best_so_far_accuracy = 0
        print("On the", i + 1, "level of the search tree")
        
        for k in range(len(data[0])):
            if (k not in current_set_of_features):
                accuracy = crossValidation(data, current_set_of_features, k, class_label, userNum)

                print("Testing adding feature", k + 1, "with current accuracy of:", accuracy, "using feature(s) { ", end = "")
                for j in current_set_of_features:
                    print(j + 1, end = ", ")
                print(k + 1, "}")

                if accuracy > best_so_far_accuracy:
                    best_so_far_accuracy = accuracy
                    feature_to_add_at_current_level = k
                if accuracy > best_total_accuracy:
                    best_total_accuracy = accuracy
                    best_feature_at_level = k
                
        print("On level", i + 1, "I added feature", feature_to_add_at_current_level+1, "with accuracy of",best_so_far_accuracy)
        current_set_of_features.append(feature_to_add_at_current_level)
        if best_feature_at_level >= 0:
            best_features.append(best_feature_at_level)
        best_feature_at_level = -1
    print("-------------------------")
    print("Best features to use:")
    for l in best_features:
        print(l + 1, end = " ")
    print("\nThe best accuracy from this data set is", best_total_accuracy)
    
    
def backwardElimination(data, class_labels, user_num):
    class_label = copy.deepcopy(class_labels)
    current_set_of_features = []
    best_features = []
    best_total_accuracy = 0.0
    accuracy = 0.0
    best_so_far_accuracy = 0.0
    
    for  i in range(len(data[0])):
        current_set_of_features.append(i)
        
    for i in range(len(data[0])): 
        feature_to_remove_at_current_level = -1
        best_so_far_accuracy = 0
        print("On the", i + 1, "level of the search tree")
        
        for k in range(len(data[0])):
            if (k in current_set_of_features):
                accuracy = crossValidation(data, current_set_of_features, k, class_label, user_num)
                print("Testing removing feature", k + 1, "with current accuracy of:", accuracy, "using feature(s) { ", end = "")
                print_features = copy.deepcopy(current_set_of_features)
                print_features.remove(k)
                for l in range(len(print_features)):
                    if l < len(print_features)-1:
                        print(print_features[l] + 1, end = ", ")
                    else:
                        print(print_features[l] + 1, end = "")
                print(" }")

                if accuracy > best_so_far_accuracy:
                    best_so_far_accuracy = accuracy
                    feature_to_remove_at_current_level = k
                if accuracy > best_total_accuracy:
                    best_total_accuracy = accuracy
                    best_features = copy.deepcopy(current_set_of_features)
                    best_features.remove(k)
                    
        print("On level", i + 1, "I removed feature", feature_to_remove_at_current_level+1)
        current_set_of_features.remove(feature_to_remove_at_current_level)
    print("-------------------------")
    print("Best features to use:")
    for l in best_features:
        print(l + 1, end = " ")
    print("\nThe best accuracy from this data set is", best_total_accuracy)
    
def main():
    print("Welcome to Nathan's Feature Selection Algorithm.")
    user_file = input("Type in the name of the file to test: ")
    print("Type the number of the algorithm you want to run.")
    user_num = input("1. Forward Selection \n2. Backward Elimination\n")
    data = np.loadtxt(user_file)
    list_data = np.ndarray.tolist(data)
    class_labels = getLabels(list_data)
    data_minus_class = editData(list_data)
    if user_num == "1":
        forwardSelection(data_minus_class, class_labels, user_num)
    elif user_num == "2":
        backwardElimination(data_minus_class, class_labels, user_num)
    else:
        print("Invalid input. Exiting")
    
if __name__ == "__main__":
    main()